core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -0,0 +1,481 @@
+      mResource = resource;
+      mLastAccessTimeMs = System.currentTimeMillis();
+    }
+  }
+
+  /**
+   * Options to initialize a Dynaminic resource pool.
+   */
+  public static class Options {
+    private int mMaxCapacity = 1024;
+    private int mMinCapacity = 1;
+    private int mInitialDelayMs = 100;
+    private int mGcIntervalMs = 120 * Constants.SECOND_MS;
+
+    /**
+     * @return the max capacity
+     */
+    public int getMaxCapacity() {
+      return mMaxCapacity;
+    }
+
+    /**
+     * @return the min capacity
+     */
+    public int getMinCapacity() {
+      return mMinCapacity;
+    }
+
+    /**
+     * @return the initial delay
+     */
+    public int getInitialDelayMs() {
+      return mInitialDelayMs;
+    }
+
+    /**
+     * @return the gc interval
+     */
+    public int getGcIntervalMs() {
+      return mGcIntervalMs;
+    }
+
+    /**
+     * @param maxCapacity the max capacity
+     * @return the current object
+     */
+    public Options setMaxCapacity(int maxCapacity) {
+      mMaxCapacity = maxCapacity;
+      return this;
+    }
+
+    /**
+     * @param minCapacity the min capacity
+     * @return the current object
+     */
+    public Options setMinCapacity(int minCapacity) {
+      mMinCapacity = minCapacity;
+      return this;
+    }
+
+    /**
+     * @param initialDelayMs the initial delay
+     * @return the current object
+     */
+    public Options setInitialDelayMs(int initialDelayMs) {
+      mInitialDelayMs = initialDelayMs;
+      return this;
+    }
+
+    /**
+     * @param gcIntervalMs the gc interval
+     * @return the current object
+     */
+    public Options setGcIntervalMs(int gcIntervalMs) {
+      mGcIntervalMs = gcIntervalMs;
+      return this;
+    }
+
+    /**
+     * Creates an option instance.
+     */
+    public Options() {
+    }
+
+    /**
+     * @return the default option
+     */
+    public static Options defaultOptions() {
+      return new Options();
+    }
+  }
+
+  private final ReentrantLock mLock = new ReentrantLock();
+  private final Condition mNotEmpty = mLock.newCondition();
+  private final int mMaxCapacity;
+  private final int mMinCapacity;
+
+  // Tracks the resources that are available ordered by lastAccessTime (the first one is
+  // the most recently used resource).
+  @GuardedBy("mLock")
+  private TreeSet<ResourceInternal<T>> mResourceAvailable =
+      new TreeSet<>(new Comparator<ResourceInternal<T>>() {
+        @Override
+        public int compare(ResourceInternal<T> c1, ResourceInternal<T> c2) {
+          if (c1 == c2) {
+            return 0;
+          }
+          if (c1.mLastAccessTimeMs == c2.mLastAccessTimeMs) {
+            return c1.mIdentity - c2.mIdentity;
+          }
+          return (int) (c2.mLastAccessTimeMs - c1.mLastAccessTimeMs);
+        }
+      });
+
+  // Tracks all the resources that are not closed.
+  @GuardedBy("mLock")
+  private HashMap<T, ResourceInternal<T>> mResources = new HashMap<>(32);
```

Comment:
Yes for some places. I started with ConcurrentHashMap then switched to this.

Basically, the lock the controls the condition has to be acquired when the condition changes. I think this is a problem (might not be a big deal in practice in most cases) in the current ResourcePool implementation.